### PR TITLE
Fix problem with null values

### PR DESCRIPTION
**Pull request by Sawadski**
@numbnut @kaihil @dbuni Please have a look at this merge request. 

-------
* Removed nodes and links if value is 0.
* Show kibana empty notification if there 0 nodes.

### CODE REVIEW

File: public/lib/validate.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  const _getNodeName = (nodes, index) => nodes[index].name;
+
+  /**
+   * Gets node index of current name.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {string} name - The node name.
+   * @returns {number} - Current node index. Returns `-1` if node name not found.
+   */
+  const _getNodeId = (nodes, name) => nodes.findIndex(node => node.name === name);
```
Comment (numbnut) [inline public/lib/validate.js]: I would propose to call this _getNodeIndex, because it is more specific. An id could be anything.

File: public/lib/validate.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+  /**
+   * Add node to node-list.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {string} name - The node name.
+   * @returns {array<NODE_OBJ>} - New list of node objects.
+   */
+  const _addNode = (nodes, name) => {
+    let id = _getNodeId(nodes, name);
```
Comment (numbnut) [inline public/lib/validate.js]: Same here: I would name the variable index.

File: public/lib/validate.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  /**
+   * Generate a hashmap.
+   * Generates a key for each link.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @returns {HASHMAP} - Hashmap of links.
+   */
+  const _generateHashMap = (nodes, links) => {
+    const data = new Map();
```
Comment (numbnut) [inline public/lib/validate.js]: I would rename data to linkMap.

File: public/lib/validate.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  /**
+   * Generate a hashmap.
+   * Generates a key for each link.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @returns {HASHMAP} - Hashmap of links.
+   */
+  const _generateHashMap = (nodes, links) => {
```
Comment (numbnut) [inline public/lib/validate.js]: I would rename the function to _generateLinkMap

File: public/lib/validate.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @returns {object} - Valid nodes and links.
+   */
+  const validate = (nodes, links) => {
+    // generate hashmap
+    const dataMap = _generateHashMap(nodes, links);
+    // delete unvalid links and nodes
+    const filteredDataMap = _filterHashMap(dataMap);
+    // split hashmap into two arrays
```
Comment (numbnut) [inline public/lib/validate.js]: I would rephrase this comment to something like
convert linkMap to link and node array

File: public/lib/validate.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+  /**
+   * Filters invalid values from hashmap.
+   * An invalid value could be a `0` value.
+   *
+   * @param {HASHMAP} data - The hashmap.
+   * @returns {HASHMAP} - Without invalid values.
+   */
+  const _filterHashMap = (data) => {
+    data.forEach((link, key) => {
```
[the remark this answers is not in the document]
Comment (Sawadski) [reply public/lib/validate.js]: A javascript `Map`-class has no filter function.
